### DISCUSSION TITLE
Slow compilation speed

### DISCUSSION THREAD

**Discussion by bizehao** [microsoft/WindowsAppSDK#2300] [Q&A] [answered]
Winui3 CPP project why sometimes compilation is particularly slow Just changed one line of code and compiled it for 5 minutes,
in win10 vs2022 v17.1.1

**Comment by castorix** [comment]
> Winui3 CPP project why sometimes compilation is particularly slow Just changed one line of code and compiled it for 5 minutes, in win10 vs2022 v17.1.1

Yes, on my OS (Windows 10 21H1, on an old PC (7 years)), all C++/WinRT projects (WinUI or not) take ages to be compiled, while the similar code in C#, VB.NET, C++/Win32 , C++/CLI is Ok

**Comment by ghost** [comment, marked as answer]
### **Console :** 

#### Win32 :         
 one (lol.cpp) file  -> compiled to (.whatever) object code file.
#### WinRT : 
**First run :** two (pch.h and lol.cpp) file -> compiled to (.whatever)  object code file. 
**Second run** : one/only changes (lol.cpp) file -> compiled to (.whatever)  object code file. 

#
 
NOTE : second run is as fast as win32 projects compile time. 
👉**The Fix :** need more mature C++ modules support in compilers.  see https://github.com/microsoft/cppwinrt/issues/847#issuecomment-770401790

----

### **XAML (WinUI 3) :**

each .XAML (WinUI 3)  files -> parsed and interpreted to ("iso standard cpp") *.cpp files ->  compiled to  (.xbf)  object code files. 

#
👉yes "compiling" WinUI 3 is a two step task and .xaml to **_"iso standard cpp"_** .cpp alone takes huge amount of time and then there's regular .cpp compilation.  see https://github.com/microsoft/cppwinrt/issues/660#issuecomment-643673750

---

and finally also see some VS crapiness https://github.com/microsoft/cppwinrt/issues/664#issuecomment-643714063

**Comment by bizehao** [comment]
ok,I suspect it is the problem of vs. sometimes the CPU is not occupied when it is very slow.
